### PR TITLE
🥡 Feat/4 Implement DB Models

### PR DESCRIPTION
**Pull request by demariadaniel**
- Adds basic PostGres / Drizzle Integration
- Establishes Models based off: https://github.com/Pan-Canadian-Genome-Library/daco/tree/main/docs/model
- Some package Type definition updates w/ UI adjusted to reflect above docs (not fully integrated w/ DB)

Note: there's extra commit history from originally basing this branch off of https://github.com/Pan-Canadian-Genome-Library/daco/pull/7
Will resolve through squash merge

### CODE REVIEW

File: apps/api/src/db/schemas/actions.ts [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+ * IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+import { relations } from 'drizzle-orm';
+import { bigint, pgEnum, pgTable, timestamp, varchar } from 'drizzle-orm/pg-core';
+import { applications } from './applications.ts';
+import { revisionRequests } from './revisionRequests.ts';
+// TODO: Integrate w/ TS
+// import { applicationActions } from 'pcgl-daco/packages/data-model/';
```
Comment (demariadaniel) [inline apps/api/src/db/schemas/actions.ts:25]: There's Type issues related to pgEnum I haven't untangled, for now this is not fully implemented

File: apps/api/src/db/schemas/applicationContents.ts [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+	institution_building: varchar({ length: 255 }),
+	// Project
+	project_title: text(),
+	project_website: text(),
+	project_abstract: text(),
+	project_methodology: text(),
+	project_summary: text(),
+	project_publication_urls: text().array(),
+	// Studies
+	// TODO: requested study information
```
Comment (demariadaniel) [inline apps/api/src/db/schemas/applicationContents.ts:68]: Related to: https://github.com/Pan-Canadian-Genome-Library/daco/blob/main/docs/model/data-model.dbml#L107

File: packages/data-model/src/types.ts [hunk 3]
```diff
@@ -44,68 +115,106 @@ export type Institution = {
...
+export type ApplicationActionData = {
+	id: number; // TODO: Implement BigInt;
+	applicationId: number; // TODO: Implement BigInt;
+	createdAt: Date;
+	userId: string;
+	action: ApplicationActions;
+	stateBefore: ApplicationStates;
+	stateAfter: ApplicationStates;
+	revisionsRequestId: number; // TODO: Implement BigInt
+	// TODO: may need reference to a content diff
```
Comment (demariadaniel) [inline packages/data-model/src/types.ts:160]: Related to: https://github.com/Pan-Canadian-Genome-Library/daco/blob/main/docs/model/data-model.dbml#L143

**Comment by ciaranschutte** [comment]
@demariadaniel should this be going into a `develop` branch vs main?

**Comment by ciaranschutte** [comment]
can remove .mts files that were renamed to .ts

File: apps/api/src/db/index.ts [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+		await testFiles();
+		await testRevisions();
+	} catch (err) {
+		console.error('Error at TestDb');
+		console.error(err);
+	}
+	process.exit();
+}
+
+await testDb();
```
Comment (ciaranschutte) [inline apps/api/src/db/index.ts:72]: unclear why all of this functionality is coded as async operations?

**Comment by demariadaniel** [comment]
> can remove .mts files that were renamed to .ts

Sorry there's some merge changes to clean up, missed that

File: apps/api/src/db/index.ts (hunk 4 above)
Comment (demariadaniel) [reply apps/api/src/db/index.ts:72]: There were original Drizzle functions that were async, I ended up making the test functions extremely basic but didn't revert the async

**Comment by demariadaniel** [comment]
> @demariadaniel should this be going into a `develop` branch vs main?

There's neither `develop` or `release` yet

File: apps/api/src/db/schemas/files.ts [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+import { relations } from 'drizzle-orm';
+import { bigint, customType, pgEnum, pgTable, timestamp, varchar } from 'drizzle-orm/pg-core';
+import { applications } from './applications.ts';
+// TODO: Integrate w/ TS
+// import { FileTypes } from 'pcgl-daco/packages/data-model/';
+
+const fileEnum = pgEnum('agreements', ['SIGNED_APPLICATION', 'ETHICS_LETTER']);
```
Comment (leoraba) [inline apps/api/src/db/schemas/files.ts:26]: enum called `agreements` already exists. guess his enum should be renamed to something like `fileTypes`

File: apps/api/src/db/schemas/applicationContents.ts [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+import { relations } from 'drizzle-orm';
+import { bigint, boolean, pgTable, text, timestamp, varchar } from 'drizzle-orm/pg-core';
+import { agreements } from './agreements.ts';
+import { applications } from './applications.ts';
+import { collaborators } from './collaborators.ts';
+import { files } from './files.ts';
+import { revisionRequests } from './revisionRequests.ts';
+
+export const applicationContents = pgTable('application_contents', {
+	id: bigint({ mode: 'number' }).primaryKey().generatedAlwaysAsIdentity(),
```
Comment (leoraba) [inline apps/api/src/db/schemas/applicationContents.ts:29]: I like you're using `bigint` and `generatedAlwaysAsIdentity` as this is the preferably way in newer versions of PostgreSQL.
Even though Drizzle docs prefers `bigserial` type (https://orm.drizzle.team/docs/column-types/pg#bigint) which looks it's outdated..

File: apps/api/src/db/schemas/applicationContents.ts [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+	applicant_profile_url: text(),
+	// Institutional Rep
+	institutional_rep_title: varchar({ length: 255 }),
+	institutional_rep_first_name: varchar({ length: 255 }),
+	institutional_rep_middle_name: varchar({ length: 255 }),
+	institutional_rep_last_name: varchar({ length: 255 }),
+	institutional_rep_suffix: varchar({ length: 255 }),
+	institutional_rep_primary_affiliation: varchar({ length: 255 }),
+	institutional_rep_email: varchar({ length: 255 }),
+	institutional_rep_profile_url: varchar({ length: 255 }),
```
Comment (leoraba) [inline apps/api/src/db/schemas/applicationContents.ts:51]: having the property `length` in here makes me think app should trim texts before inserting. or db could throw an error on larger texts.

File: apps/api/src/db/schemas/revisionRequests.ts [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+
+export const revisionRequests = pgTable('revisionRequests', {
+	id: bigint({ mode: 'number' }).primaryKey().generatedAlwaysAsIdentity(),
+	application_id: bigint({ mode: 'number' }).notNull(),
+	created_at: timestamp().notNull(),
+	comments: text(),
+	applicant_notes: text(),
+	applicant_approved: boolean().notNull(),
+	institution_rep_approved: boolean().notNull(),
+	institution_rep_notes: boolean().notNull(),
```
Comment (leoraba) [inline apps/api/src/db/schemas/revisionRequests.ts:33]: `institution_rep_notes` should be a text() based on the DBML

**Review by leoraba — COMMENTED** [review]
👍 just added some comments.

File: apps/api/src/db/schemas/files.ts (hunk 5 above)
Comment (demariadaniel) [reply apps/api/src/db/schemas/files.ts]: Oops, thanks. Updated enum & variable name to use `fileTypes`

File: apps/api/src/db/schemas/applicationContents.ts (hunk 7 above)
Comment (demariadaniel) [reply apps/api/src/db/schemas/applicationContents.ts:51]: UI / API validation is still down the road. DBML specifies length for these fields: https://github.com/Pan-Canadian-Genome-Library/daco/blob/main/docs/model/data-model.dbml#L91

File: apps/api/src/db/schemas/revisionRequests.ts (hunk 8 above)
Comment (demariadaniel) [reply apps/api/src/db/schemas/revisionRequests.ts]: Oh thanks, could possibly be null as well

**Comment by justincorrigible** [comment]
Looks like this PR got merged before approvals? 🤔 
think I may have missed placing branch protections... apologies. let me fix that right now!

**Pull request merged**